src/js/actions/mask.js

```diff
@@ -330,6 +334,88 @@ define(function (require, exports) {
         transfers: ["tools.select"]
     };
 
+    /**
+     * if there are two selected layers create a vector mask on one of them from the path of the shape in the 
+     * other selected layer
+     *
+     * @return {Promise}
+     */
+    var createVectorMaskFromShape = function () {
+        var appStore = this.flux.store("application"),
+            currentDocument = appStore.getCurrentDocument();
+
+        if (!currentDocument) {
+            return Promise.resolve();
+        }
+
+        var currentLayers = currentDocument.layers.selected;
+
+        if (currentLayers.size !== 2) {
+            return Promise.resolve();
+        }
+
+        var shapeLayers = currentLayers.filter(function (layer) {
+                return layer.isVector;
+            }),
+            maskLayers = currentLayers.filter(function (layer) {
+                return currentDocument.layers.canSupportVectorMask(layer) &&
+                    !layer.vectorMaskEnabled;
+            });
+
+        if (shapeLayers.isEmpty() || maskLayers.isEmpty()) {
+            return Promise.resolve();
+        }
+       
+        var shapeLayer = shapeLayers.first(),
+            maskLayer = maskLayers.first(),
+            shapeLayerRef = layerLib.referenceBy.id(shapeLayer.id);
+
+        var options = {
+            paintOptions: {
+                immediateUpdate: true,
+                quality: "draft"
+            },
+            historyStateInfo: {
+                name: nls.localize("strings.ACTIONS.ADD_VECTOR_MASK"),
+                target: documentLib.referenceBy.id(currentDocument.id),
+                coalesce: false,
+                suppressHistoryStateNotification: false
+            }
+        },
+            playObjects = [layerLib.select(shapeLayerRef),
+                vectorMaskLib.createMaskFromShape(maskLayer.id)],
+            notShapeLayer = collection.difference(currentDocument.layers.all,
+                Immutable.List.of(shapeLayer)),
+            notShapeLayerContainsLeaf = notShapeLayer.some(function (layer) {
+                    return !layer.isGroupEnd && !layer.isGroup;
+                }),
+            deleteShapeLayer = !notShapeLayer.isEmpty() && notShapeLayerContainsLeaf;
```

Comment:
It seems like this could just as easily be computed without creating an intermediate data structure with `collection.difference` that contains all but one layer by just adding one more conjunct to to the `.some` call above.